<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Peter von Rohr">
<meta name="dcterms.date" content="2024-04-24">

<title>psbss2024 | 0.0.4 - Model Selection and Variance Components</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">psbss2024 | 0.0.4</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../course_notes.html"> 
<span class="menu-text">Course Notes</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../exercises.html"> 
<span class="menu-text">Exercises</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../solutions.html"> 
<span class="menu-text">Solutions</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../add_info.qmd"> 
<span class="menu-text">Additional Information</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../News.html"> 
<span class="menu-text">News</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#breeding-programs" id="toc-breeding-programs" class="nav-link active" data-scroll-target="#breeding-programs">Breeding Programs</a></li>
  <li><a href="#new-trait" id="toc-new-trait" class="nav-link" data-scroll-target="#new-trait">New Trait</a></li>
  <li><a href="#background-and-context" id="toc-background-and-context" class="nav-link" data-scroll-target="#background-and-context">Background and Context</a></li>
  <li><a href="#genetic-improvement" id="toc-genetic-improvement" class="nav-link" data-scroll-target="#genetic-improvement">Genetic Improvement</a></li>
  <li><a href="#why-statistical-modelling" id="toc-why-statistical-modelling" class="nav-link" data-scroll-target="#why-statistical-modelling">Why Statistical Modelling?</a></li>
  <li><a href="#statistical-modelling-because" id="toc-statistical-modelling-because" class="nav-link" data-scroll-target="#statistical-modelling-because">Statistical Modelling Because …</a></li>
  <li><a href="#deterministic-versus-stochastic" id="toc-deterministic-versus-stochastic" class="nav-link" data-scroll-target="#deterministic-versus-stochastic">Deterministic Versus Stochastic</a></li>
  <li><a href="#statistical-model" id="toc-statistical-model" class="nav-link" data-scroll-target="#statistical-model">Statistical Model</a></li>
  <li><a href="#how-does-a-statistical-model-work" id="toc-how-does-a-statistical-model-work" class="nav-link" data-scroll-target="#how-does-a-statistical-model-work">How Does A Statistical Model Work?</a></li>
  <li><a href="#which-function-mx" id="toc-which-function-mx" class="nav-link" data-scroll-target="#which-function-mx">Which function <span class="math inline">\(m(x)\)</span>?</a></li>
  <li><a href="#which-predictor-variables" id="toc-which-predictor-variables" class="nav-link" data-scroll-target="#which-predictor-variables">Which predictor variables?</a></li>
  <li><a href="#why-model-selection" id="toc-why-model-selection" class="nav-link" data-scroll-target="#why-model-selection">Why Model Selection</a></li>
  <li><a href="#example-dataset" id="toc-example-dataset" class="nav-link" data-scroll-target="#example-dataset">Example Dataset</a></li>
  <li><a href="#no-relevance-of-predictors" id="toc-no-relevance-of-predictors" class="nav-link" data-scroll-target="#no-relevance-of-predictors">No Relevance of Predictors</a></li>
  <li><a href="#relevance-of-predictors" id="toc-relevance-of-predictors" class="nav-link" data-scroll-target="#relevance-of-predictors">Relevance of Predictors</a></li>
  <li><a href="#fitting-a-regression-model" id="toc-fitting-a-regression-model" class="nav-link" data-scroll-target="#fitting-a-regression-model">Fitting a Regression Model</a></li>
  <li><a href="#fitting-a-regression-model-ii" id="toc-fitting-a-regression-model-ii" class="nav-link" data-scroll-target="#fitting-a-regression-model-ii">Fitting a Regression Model II</a></li>
  <li><a href="#multiple-regression" id="toc-multiple-regression" class="nav-link" data-scroll-target="#multiple-regression">Multiple Regression</a></li>
  <li><a href="#which-model-is-better" id="toc-which-model-is-better" class="nav-link" data-scroll-target="#which-model-is-better">Which model is better?</a></li>
  <li><a href="#bias-variance-trade-off" id="toc-bias-variance-trade-off" class="nav-link" data-scroll-target="#bias-variance-trade-off">Bias-variance trade-off</a></li>
  <li><a href="#bias-variance-trade-off-ii" id="toc-bias-variance-trade-off-ii" class="nav-link" data-scroll-target="#bias-variance-trade-off-ii">Bias-variance trade-off II</a></li>
  <li><a href="#mallows-c_p-statistic" id="toc-mallows-c_p-statistic" class="nav-link" data-scroll-target="#mallows-c_p-statistic">Mallows <span class="math inline">\(C_p\)</span> statistic</a></li>
  <li><a href="#searching-the-best-model" id="toc-searching-the-best-model" class="nav-link" data-scroll-target="#searching-the-best-model">Searching The Best Model</a></li>
  <li><a href="#forward-selection" id="toc-forward-selection" class="nav-link" data-scroll-target="#forward-selection">Forward Selection</a></li>
  <li><a href="#backward-selection" id="toc-backward-selection" class="nav-link" data-scroll-target="#backward-selection">Backward Selection</a></li>
  <li><a href="#considerations" id="toc-considerations" class="nav-link" data-scroll-target="#considerations">Considerations</a></li>
  <li><a href="#alternative-selection-criteria" id="toc-alternative-selection-criteria" class="nav-link" data-scroll-target="#alternative-selection-criteria">Alternative Selection Criteria</a></li>
  <li><a href="#genetic-variation" id="toc-genetic-variation" class="nav-link" data-scroll-target="#genetic-variation">Genetic Variation</a></li>
  <li><a href="#two-traits" id="toc-two-traits" class="nav-link" data-scroll-target="#two-traits">Two Traits</a></li>
  <li><a href="#problems" id="toc-problems" class="nav-link" data-scroll-target="#problems">Problems</a></li>
  <li><a href="#example-of-variance-components-separation" id="toc-example-of-variance-components-separation" class="nav-link" data-scroll-target="#example-of-variance-components-separation">Example of Variance Components Separation</a></li>
  <li><a href="#repeatability-plot" id="toc-repeatability-plot" class="nav-link" data-scroll-target="#repeatability-plot">Repeatability Plot</a></li>
  <li><a href="#model" id="toc-model" class="nav-link" data-scroll-target="#model">Model</a></li>
  <li><a href="#estimation-of-variance-components" id="toc-estimation-of-variance-components" class="nav-link" data-scroll-target="#estimation-of-variance-components">Estimation Of Variance Components</a></li>
  <li><a href="#analysis-of-variance-anova" id="toc-analysis-of-variance-anova" class="nav-link" data-scroll-target="#analysis-of-variance-anova">Analysis Of Variance (ANOVA)</a></li>
  <li><a href="#zahlenbeispiel" id="toc-zahlenbeispiel" class="nav-link" data-scroll-target="#zahlenbeispiel">Zahlenbeispiel</a></li>
  <li><a href="#same-strategy-for-sire-model" id="toc-same-strategy-for-sire-model" class="nav-link" data-scroll-target="#same-strategy-for-sire-model">Same Strategy for Sire Model</a></li>
  <li><a href="#anova" id="toc-anova" class="nav-link" data-scroll-target="#anova">ANOVA</a></li>
  <li><a href="#maximum-likelihood-ml" id="toc-maximum-likelihood-ml" class="nav-link" data-scroll-target="#maximum-likelihood-ml">Maximum Likelihood (ML)</a></li>
  <li><a href="#maximization-of-likelihood" id="toc-maximization-of-likelihood" class="nav-link" data-scroll-target="#maximization-of-likelihood">Maximization of Likelihood</a></li>
  <li><a href="#restricted-maximum-likelihood-reml" id="toc-restricted-maximum-likelihood-reml" class="nav-link" data-scroll-target="#restricted-maximum-likelihood-reml">Restricted Maximum Likelihood (REML)</a></li>
  <li><a href="#reml-ii" id="toc-reml-ii" class="nav-link" data-scroll-target="#reml-ii">REML II</a></li>
  <li><a href="#bayesian-estimation" id="toc-bayesian-estimation" class="nav-link" data-scroll-target="#bayesian-estimation">Bayesian Estimation</a></li>
  <li><a href="#model-1" id="toc-model-1" class="nav-link" data-scroll-target="#model-1">Model</a></li>
  <li><a href="#likelihood" id="toc-likelihood" class="nav-link" data-scroll-target="#likelihood">Likelihood</a></li>
  <li><a href="#priors" id="toc-priors" class="nav-link" data-scroll-target="#priors">Priors</a></li>
  <li><a href="#additional-terms" id="toc-additional-terms" class="nav-link" data-scroll-target="#additional-terms">Additional Terms</a></li>
  <li><a href="#joint-posterior-density" id="toc-joint-posterior-density" class="nav-link" data-scroll-target="#joint-posterior-density">Joint Posterior Density</a></li>
  <li><a href="#fully-conditional-posterior-densities-of-theta" id="toc-fully-conditional-posterior-densities-of-theta" class="nav-link" data-scroll-target="#fully-conditional-posterior-densities-of-theta">Fully Conditional Posterior Densities of <span class="math inline">\(\theta\)</span></a></li>
  <li><a href="#fully-conditional-posterior-densities-of-sigma_e2" id="toc-fully-conditional-posterior-densities-of-sigma_e2" class="nav-link" data-scroll-target="#fully-conditional-posterior-densities-of-sigma_e2">Fully Conditional Posterior Densities of <span class="math inline">\(\sigma_e^2\)</span></a></li>
  <li><a href="#fully-conditional-posterior-densities-of-sigma_u2" id="toc-fully-conditional-posterior-densities-of-sigma_u2" class="nav-link" data-scroll-target="#fully-conditional-posterior-densities-of-sigma_u2">Fully Conditional Posterior Densities of <span class="math inline">\(\sigma_u^2\)</span></a></li>
  <li><a href="#implementation" id="toc-implementation" class="nav-link" data-scroll-target="#implementation">Implementation</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Model Selection and Variance Components</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Peter von Rohr </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">April 24, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="breeding-programs" class="level2">
<h2 class="anchored" data-anchor-id="breeding-programs">Breeding Programs</h2>
<div class="cell" data-hook_convert_odg="true" data-fig_path="odg">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="odg/bprogdiag.png" class="img-fluid figure-img" style="width:11cm" data-fig-pos="H"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="new-trait" class="level2">
<h2 class="anchored" data-anchor-id="new-trait">New Trait</h2>
<ul>
<li>New trait to be considered in breeding program</li>
<li>Why? <span class="math inline">\(\rightarrow\)</span> Trait is of economic importance</li>
<li>Want to improve average level of trait in a given population</li>
<li>How is this done?</li>
<li>What do we have to do?</li>
</ul>
</section>
<section id="background-and-context" class="level2">
<h2 class="anchored" data-anchor-id="background-and-context">Background and Context</h2>
<ul>
<li>Farms/Enterprise use livestock products as base for economic existence</li>
<li>Improvements of production efficiency improves sustainability</li>
<li>Short-term:
<ul>
<li>improve management and environment</li>
<li>select optimal livestock breed / population for given environment</li>
</ul></li>
<li>Long-term:
<ul>
<li>improve population at genetic level</li>
<li>define breeding goal</li>
<li>select parents such that offspring are “closer” to goal compared to parents</li>
</ul></li>
</ul>
</section>
<section id="genetic-improvement" class="level2">
<h2 class="anchored" data-anchor-id="genetic-improvement">Genetic Improvement</h2>
<ul>
<li>Genetic improvement happens between parents and offspring</li>
<li>Parents pass random sample of alleles to offspring</li>
<li>Goal: select parents that have many “good” alleles to pass to offspring</li>
<li>How to find parents with “good” alleles without knowing which genes are important?</li>
</ul>
<p><span class="math inline">\(\rightarrow\)</span> <strong>Statistical Modeling</strong></p>
</section>
<section id="why-statistical-modelling" class="level2">
<h2 class="anchored" data-anchor-id="why-statistical-modelling">Why Statistical Modelling?</h2>
<p>Some people believe, they do not need statistics. For them it is enough to look at a diagram</p>
<div class="cell" data-hook_convert_odg="true" data-fig_path="odg">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="odg/no-stat-model-diagram.png" style="height:7cm" width="284" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="statistical-modelling-because" class="level2">
<h2 class="anchored" data-anchor-id="statistical-modelling-because">Statistical Modelling Because …</h2>
<p>Two types of dependencies between physical quantities</p>
<ol type="1">
<li>deterministic</li>
<li>stochastic</li>
</ol>
</section>
<section id="deterministic-versus-stochastic" class="level2">
<h2 class="anchored" data-anchor-id="deterministic-versus-stochastic">Deterministic Versus Stochastic</h2>
<div class="cell" data-hook_convert_odg="true" data-fig_path="odg">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="odg/deter-vs-stoch.png" class="img-fluid figure-img" width="381"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="statistical-model" class="level2">
<h2 class="anchored" data-anchor-id="statistical-model">Statistical Model</h2>
<ul>
<li>stochastic systems contains many sources of uncertainty</li>
<li>statistical models can handle uncertainty</li>
<li>components of a statistical model
<ul>
<li>response variable <span class="math inline">\(y\)</span></li>
<li>predictor variables <span class="math inline">\(x_1, x_2, \ldots, x_k\)</span></li>
<li>error term <span class="math inline">\(e\)</span></li>
<li>function <span class="math inline">\(m(x)\)</span></li>
</ul></li>
</ul>
</section>
<section id="how-does-a-statistical-model-work" class="level2">
<h2 class="anchored" data-anchor-id="how-does-a-statistical-model-work">How Does A Statistical Model Work?</h2>
<ul>
<li>predictor variables <span class="math inline">\(x_1, x_2, \ldots, x_k\)</span> are transformed by function <span class="math inline">\(m(x)\)</span> to explain the response variable <span class="math inline">\(y\)</span></li>
<li>uncertainty is captured by error term.</li>
<li>as a formula, for observation <span class="math inline">\(i\)</span></li>
</ul>
<p><span class="math display">\[y_i = m(x_i) + e_i\]</span></p>
</section>
<section id="which-function-mx" class="level2">
<h2 class="anchored" data-anchor-id="which-function-mx">Which function <span class="math inline">\(m(x)\)</span>?</h2>
<ul>
<li>class of functions that can be used as <span class="math inline">\(m(x)\)</span> is infinitely large</li>
<li>restrict to linear functions of predictor variables</li>
</ul>
</section>
<section id="which-predictor-variables" class="level2">
<h2 class="anchored" data-anchor-id="which-predictor-variables">Which predictor variables?</h2>
<ul>
<li>Question, about which predictor variables to use is answered by model selection</li>
</ul>
</section>
<section id="why-model-selection" class="level2">
<h2 class="anchored" data-anchor-id="why-model-selection">Why Model Selection</h2>
<ul>
<li>Many predictor variables are available</li>
<li>Are all of them <code>relevant</code>?</li>
<li>What is the meaning of <code>relevant</code> in this context?</li>
</ul>
</section>
<section id="example-dataset" class="level2">
<h2 class="anchored" data-anchor-id="example-dataset">Example Dataset</h2>
<div class="cell">
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: right;">Animal</th>
<th style="text-align: right;">Breast Circumference</th>
<th style="text-align: right;">Body Weight</th>
<th style="text-align: right;">RandPred</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: right;">1</td>
<td style="text-align: right;">176</td>
<td style="text-align: right;">471</td>
<td style="text-align: right;">177</td>
</tr>
<tr class="even">
<td style="text-align: right;">2</td>
<td style="text-align: right;">177</td>
<td style="text-align: right;">463</td>
<td style="text-align: right;">177</td>
</tr>
<tr class="odd">
<td style="text-align: right;">3</td>
<td style="text-align: right;">178</td>
<td style="text-align: right;">481</td>
<td style="text-align: right;">180</td>
</tr>
<tr class="even">
<td style="text-align: right;">4</td>
<td style="text-align: right;">179</td>
<td style="text-align: right;">470</td>
<td style="text-align: right;">178</td>
</tr>
<tr class="odd">
<td style="text-align: right;">5</td>
<td style="text-align: right;">179</td>
<td style="text-align: right;">496</td>
<td style="text-align: right;">182</td>
</tr>
<tr class="even">
<td style="text-align: right;">6</td>
<td style="text-align: right;">180</td>
<td style="text-align: right;">491</td>
<td style="text-align: right;">179</td>
</tr>
<tr class="odd">
<td style="text-align: right;">7</td>
<td style="text-align: right;">181</td>
<td style="text-align: right;">518</td>
<td style="text-align: right;">180</td>
</tr>
<tr class="even">
<td style="text-align: right;">8</td>
<td style="text-align: right;">182</td>
<td style="text-align: right;">511</td>
<td style="text-align: right;">184</td>
</tr>
<tr class="odd">
<td style="text-align: right;">9</td>
<td style="text-align: right;">183</td>
<td style="text-align: right;">510</td>
<td style="text-align: right;">179</td>
</tr>
<tr class="even">
<td style="text-align: right;">10</td>
<td style="text-align: right;">184</td>
<td style="text-align: right;">541</td>
<td style="text-align: right;">176</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="no-relevance-of-predictors" class="level2">
<h2 class="anchored" data-anchor-id="no-relevance-of-predictors">No Relevance of Predictors</h2>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="l02_psb_modsel_varcomp_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="relevance-of-predictors" class="level2">
<h2 class="anchored" data-anchor-id="relevance-of-predictors">Relevance of Predictors</h2>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="l02_psb_modsel_varcomp_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="fitting-a-regression-model" class="level2">
<h2 class="anchored" data-anchor-id="fitting-a-regression-model">Fitting a Regression Model</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = `Body Weight` ~ RandPred, data = tbl_reg_aug)

Residuals:
    Min      1Q  Median      3Q     Max 
-28.481 -19.249  -3.897  13.275  51.209 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)
(Intercept)  192.299    636.596   0.302    0.770
RandPred       1.690      3.552   0.476    0.647

Residual standard error: 26.01 on 8 degrees of freedom
Multiple R-squared:  0.02753,   Adjusted R-squared:  -0.09403 
F-statistic: 0.2264 on 1 and 8 DF,  p-value: 0.6469</code></pre>
</div>
</div>
</section>
<section id="fitting-a-regression-model-ii" class="level2">
<h2 class="anchored" data-anchor-id="fitting-a-regression-model-ii">Fitting a Regression Model II</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = `Body Weight` ~ `Breast Circumference`, data = tbl_reg_aug)

Residuals:
     Min       1Q   Median       3Q      Max 
-17.3941  -6.5525  -0.0673   9.3707  13.2594 

Coefficients:
                        Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)            -1065.115    255.483  -4.169 0.003126 ** 
`Breast Circumference`     8.673      1.420   6.108 0.000287 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 11.08 on 8 degrees of freedom
Multiple R-squared:  0.8234,    Adjusted R-squared:  0.8014 
F-statistic: 37.31 on 1 and 8 DF,  p-value: 0.000287</code></pre>
</div>
</div>
</section>
<section id="multiple-regression" class="level2">
<h2 class="anchored" data-anchor-id="multiple-regression">Multiple Regression</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = `Body Weight` ~ `Breast Circumference` + RandPred, 
    data = tbl_reg_aug)

Residuals:
     Min       1Q   Median       3Q      Max 
-17.1739  -6.2679  -0.6437   9.4688  13.1283 

Coefficients:
                        Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)            -1096.580    369.334  -2.969 0.020835 *  
`Breast Circumference`     8.642      1.536   5.625 0.000795 ***
RandPred                   0.207      1.638   0.126 0.902980    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 11.83 on 7 degrees of freedom
Multiple R-squared:  0.8238,    Adjusted R-squared:  0.7735 
F-statistic: 16.37 on 2 and 7 DF,  p-value: 0.002295</code></pre>
</div>
</div>
</section>
<section id="which-model-is-better" class="level2">
<h2 class="anchored" data-anchor-id="which-model-is-better">Which model is better?</h2>
<p>Why not taking all predictors?</p>
<ul>
<li>Additional parameters must be estimated from data</li>
<li>Predictive power decreased with too many predictors (cannot be shown for this data set, because too few data points)</li>
<li>Bias-variance trade-off</li>
</ul>
</section>
<section id="bias-variance-trade-off" class="level2">
<h2 class="anchored" data-anchor-id="bias-variance-trade-off">Bias-variance trade-off</h2>
<ul>
<li>Assume, we are looking for optimum prediction</li>
</ul>
<p><span class="math display">\[s_i = \sum_{r=1}^q \hat{\beta}_{j_r}x_{ij_r}\]</span> with <span class="math inline">\(q\)</span> relevant predictor variables</p>
<ul>
<li>Average mean squared error of prediction <span class="math inline">\(s_i\)</span></li>
</ul>
<p><span class="math display">\[MSE = n^{-1}\sum_{i=1}^n E\left[(m(x_i) - s_i)^2 \right]\]</span> where <span class="math inline">\(m(.)\)</span> denotes the linear function of the unknown true model.</p>
</section>
<section id="bias-variance-trade-off-ii" class="level2">
<h2 class="anchored" data-anchor-id="bias-variance-trade-off-ii">Bias-variance trade-off II</h2>
<ul>
<li>MSE can be split into two parts</li>
</ul>
<p><span class="math display">\[MSE = n^{-1}\sum_{i=1}^n \left( E\left[s_i \right] - m(x_i) \right)^2
  + n^{-1}\sum_{i=1}^n var(s_i)
\]</span></p>
<p>where <span class="math inline">\(n^{-1}\sum_{i=1}^n \left( E\left[s_i \right] - m(x_i) \right)^2\)</span> is called the squared <strong>bias</strong></p>
<ul>
<li>Increasing <span class="math inline">\(q\)</span> leads to reduced bias but increased variance (<span class="math inline">\(var(s_i)\)</span>)</li>
<li>Hence, find <span class="math inline">\(s_i\)</span> such that MSE is minimal</li>
<li>Problem: cannot compute MSE because <span class="math inline">\(m(.)\)</span> is not known</li>
</ul>
<p><span class="math inline">\(\rightarrow\)</span> estimate MSE</p>
</section>
<section id="mallows-c_p-statistic" class="level2">
<h2 class="anchored" data-anchor-id="mallows-c_p-statistic">Mallows <span class="math inline">\(C_p\)</span> statistic</h2>
<ul>
<li>For a given model <span class="math inline">\(\mathcal{M}\)</span>, <span class="math inline">\(SSE(\mathcal{M})\)</span> stands for the residual sum of squares.</li>
<li>MSE can be estimated as</li>
</ul>
<p><span class="math display">\[\widehat{MSE} = n^{-1} SSE(\mathcal{M}) - \hat{\sigma}^2 + 2 \hat{\sigma}^2 |\mathcal{M}|/n\]</span></p>
<p>where <span class="math inline">\(\hat{\sigma}^2\)</span> is the estimate of the error variance of the full model, <span class="math inline">\(SSE(\mathcal{M})\)</span> is the residual sum of squares of the model <span class="math inline">\(\mathcal{M}\)</span>, <span class="math inline">\(n\)</span> is the number of observations and <span class="math inline">\(|\mathcal{M}|\)</span> stands for the number of predictors in <span class="math inline">\(\mathcal{M}\)</span></p>
<p><span class="math display">\[C_p(\mathcal{M}) = \frac{SSE(\mathcal{M})}{\hat{\sigma}^2} - n + 2 |\mathcal{M}|\]</span></p>
</section>
<section id="searching-the-best-model" class="level2">
<h2 class="anchored" data-anchor-id="searching-the-best-model">Searching The Best Model</h2>
<ul>
<li>Exhaustive search over all sub-models might be too expensive</li>
<li>For <span class="math inline">\(p\)</span> predictors there are <span class="math inline">\(2^p - 1\)</span> sub-models</li>
<li>With <span class="math inline">\(p=16\)</span>, we get <span class="math inline">\(6.5535\times 10^{4}\)</span> sub-models</li>
</ul>
<p><span class="math inline">\(\rightarrow\)</span> step-wise approaches</p>
</section>
<section id="forward-selection" class="level2">
<h2 class="anchored" data-anchor-id="forward-selection">Forward Selection</h2>
<ol type="1">
<li>Start with smallest sub-model <span class="math inline">\(\mathcal{M}_0\)</span> as current model</li>
<li>Include predictor that reduces SSE the most to current model</li>
<li>Repeat step 2 until all predictors are chosen</li>
</ol>
<p><span class="math inline">\(\rightarrow\)</span> results in sequence <span class="math inline">\(\mathcal{M}_0 \subseteq \mathcal{M}_1 \subseteq \mathcal{M}_2 \subseteq \ldots\)</span> of sub-models</p>
<ol start="4" type="1">
<li>Out of sequence of sub-models choose the one with minimal <span class="math inline">\(C_p\)</span></li>
</ol>
</section>
<section id="backward-selection" class="level2">
<h2 class="anchored" data-anchor-id="backward-selection">Backward Selection</h2>
<ol type="1">
<li>Start with full model <span class="math inline">\(\mathcal{M}_0\)</span> as the current model</li>
<li>Exclude predictor variable that increases SSE the least from current model</li>
<li>Repeat step 2 until all predictors are excluded (except for intercept)</li>
</ol>
<p><span class="math inline">\(\rightarrow\)</span> results in sequence <span class="math inline">\(\mathcal{M}_0 \supseteq \mathcal{M}_1 \supseteq \mathcal{M}_2 \supseteq \ldots\)</span> of sub-models</p>
<ol start="4" type="1">
<li>Out of sequence choose the one with minimal <span class="math inline">\(C_p\)</span></li>
</ol>
</section>
<section id="considerations" class="level2">
<h2 class="anchored" data-anchor-id="considerations">Considerations</h2>
<ul>
<li>Whenever possible, choose <strong>backward</strong> selection, because it leads to better results</li>
<li>If <span class="math inline">\(p \geq n\)</span>, only forward is possible, but then consider LASSO</li>
</ul>
</section>
<section id="alternative-selection-criteria" class="level2">
<h2 class="anchored" data-anchor-id="alternative-selection-criteria">Alternative Selection Criteria</h2>
<ul>
<li>AIC or BIC, requires distributional assumptions.</li>
<li>AIC is implemented in <code>MASS::stepAIC()</code></li>
<li>Adjusted <span class="math inline">\(R^2\)</span> is a measure of goodness of fit, but sometimes is not conclusive when comparing two models</li>
<li>Try in exercise</li>
</ul>
</section>
<section id="genetic-variation" class="level2">
<h2 class="anchored" data-anchor-id="genetic-variation">Genetic Variation</h2>
<ul>
<li>Requirement for trait to be considered in breeding goal</li>
<li>Breeding means improvement of next generation via selection and mating</li>
<li>Only genetic (additive) components are passed to offspring</li>
<li>Selection should be based on genetic component of trait</li>
<li>Selection only possible with genetic variation</li>
</ul>
<p><span class="math inline">\(\rightarrow\)</span> genetic variation indicates how good characteristics are passed from parents to offspring</p>
<p><span class="math inline">\(\rightarrow\)</span> measured by <strong>heritability</strong> <span class="math inline">\(h^2 = \frac{\sigma_a^2}{\sigma_p^2}\)</span></p>
</section>
<section id="two-traits" class="level2">
<h2 class="anchored" data-anchor-id="two-traits">Two Traits</h2>
<div class="cell" data-hook_convert_odg="true" data-fig_path="odg">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="odg/twotraitwithwoutgv.png" class="img-fluid figure-img" style="width:10cm"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="problems" class="level2">
<h2 class="anchored" data-anchor-id="problems">Problems</h2>
<ul>
<li>Genetic components cannot be observed or measured</li>
<li>Must be estimated from data</li>
<li>Data are mostly phenotypic</li>
</ul>
<p><span class="math inline">\(\rightarrow\)</span> topic of variance components estimation</p>
<ul>
<li>Model based, that means connection between phenotypic measure and genetic component are based on certain model</li>
</ul>
<p><span class="math display">\[p = g + e\]</span></p>
<p>with <span class="math inline">\(cov(g,e) = 0\)</span></p>
<ul>
<li><strong>Goal</strong>: separate variation due to <span class="math inline">\(g\)</span> (<span class="math inline">\(\sigma_a^2\)</span>) from phenotypic variation</li>
</ul>
</section>
<section id="example-of-variance-components-separation" class="level2">
<h2 class="anchored" data-anchor-id="example-of-variance-components-separation">Example of Variance Components Separation</h2>
<ul>
<li>Estimation of repeatability</li>
<li>Given repeated measurements of same trait at the same animal</li>
<li>Repeatability means variation of measurements at the same animal is smaller than variation between measurements at different animals</li>
</ul>
</section>
<section id="repeatability-plot" class="level2">
<h2 class="anchored" data-anchor-id="repeatability-plot">Repeatability Plot</h2>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="l02_psb_modsel_varcomp_files/figure-html/repmeasurebulldef-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="model" class="level2">
<h2 class="anchored" data-anchor-id="model">Model</h2>
<p><span class="math display">\[\begin{equation}
y_{ij} = \mu + t_i + \epsilon_{ij} \notag
\end{equation}\]</span></p>
</section>
<section id="estimation-of-variance-components" class="level2">
<h2 class="anchored" data-anchor-id="estimation-of-variance-components">Estimation Of Variance Components</h2>
<ul>
<li><p><span class="math inline">\(E(t_i) = 0\)</span></p></li>
<li><p><span class="math inline">\(\sigma_t^2 = E(t_i^2)\)</span>: variance component of total variance (<span class="math inline">\(\sigma_y^2\)</span>) which can be attributed to the <span class="math inline">\(t\)</span>-effects</p></li>
<li><p><span class="math inline">\(E(\epsilon_{ij}) = 0\)</span></p></li>
<li><p><span class="math inline">\(\sigma_{\epsilon}^2 = E(\epsilon_{ij}^2)\)</span>: variance component attributed to <span class="math inline">\(\epsilon\)</span>-effects</p></li>
<li><p><span class="math inline">\(\sigma_y^2 = \sigma_t^2 + \sigma_{\epsilon}^2\)</span></p></li>
<li><p>Repeatability <span class="math inline">\(w\)</span> defined as:</p></li>
</ul>
<p><span class="math display">\[\begin{equation}
w = \frac{\sigma_t^2}{\sigma_t^2 + \sigma_{\epsilon}^2} \notag
\end{equation}\]</span></p>
<p><span class="math inline">\(\rightarrow\)</span> estimate of <span class="math inline">\(\sigma_t^2\)</span> needed</p>
</section>
<section id="analysis-of-variance-anova" class="level2">
<h2 class="anchored" data-anchor-id="analysis-of-variance-anova">Analysis Of Variance (ANOVA)</h2>
<p>where</p>
<!-- Essl1987 p. 212 -->
<p><span class="math display">\[SSQ(t) = \left[ {1 \over n} \sum_{i=1}^r \left( \sum_{j=1}^n y_{ij}\right)^2 \right] - \left(\sum_{i=1}^r\sum_{j=1}^{n}y_{ij}\right)^2 / N\]</span></p>
<p><span class="math display">\[SSQ(\epsilon) =  \sum_{i=1}^r\sum_{j=1}^{n}y_{ij}^2 - \left[ {1 \over n} \sum_{i=1}^r \left( \sum_{j=1}^n y_{ij}\right)^2 \right]\]</span></p>
</section>
<section id="zahlenbeispiel" class="level2">
<h2 class="anchored" data-anchor-id="zahlenbeispiel">Zahlenbeispiel</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>            Df Sum Sq Mean Sq F value   Pr(&gt;F)    
Bull         9  286.7   31.85   13.85 8.74e-07 ***
Residuals   20   46.0    2.30                     
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>Setting expected values of <code>Mean Sq</code> equal to estimates of variance components</p>
<p><span class="math display">\[\hat{\sigma}_{\epsilon}^2 = 2.3 \text{ and } \hat{\sigma}_t^2 = \frac{31.85 - 2.3}{3} = 9.85\]</span></p>
<p>Repeatability</p>
<p><span class="math display">\[\hat{w} = \frac{\hat{\sigma}_t^2}{\hat{\sigma}_t^2 + \hat{\sigma}_{\epsilon}^2} = 0.81\]</span></p>
</section>
<section id="same-strategy-for-sire-model" class="level2">
<h2 class="anchored" data-anchor-id="same-strategy-for-sire-model">Same Strategy for Sire Model</h2>
<ul>
<li>Sire model is a mixed linear effects model with sire effects <span class="math inline">\(s\)</span> as random components</li>
</ul>
<p><span class="math display">\[\begin{equation}
y = Xb + Zs + e \notag
\end{equation}\]</span></p>
<ul>
<li>In case where sires are not related, $var(s) = <span class="math inline">\(I * \sigma_s^2\)</span></li>
<li>From <span class="math inline">\(\sigma_s^2\)</span>, we get genetic additive variance as <span class="math inline">\(\sigma_a^2 = 4*\sigma_s^2\)</span></li>
</ul>
</section>
<section id="anova" class="level2">
<h2 class="anchored" data-anchor-id="anova">ANOVA</h2>
<p>with <span class="math display">\[k = \frac{1}{r-1} \left[ N - \frac{\sum_{i=1}^{r}n_i^2}{N} \right]\]</span></p>
</section>
<section id="maximum-likelihood-ml" class="level2">
<h2 class="anchored" data-anchor-id="maximum-likelihood-ml">Maximum Likelihood (ML)</h2>
<ul>
<li>Likelihood</li>
</ul>
<p><span class="math display">\[\begin{equation}
L(\theta) = f(y | \theta) \notag
\end{equation}\]</span></p>
<ul>
<li>Normal distribution</li>
</ul>
<p><span class="math display">\[\begin{equation}
L(\theta) = (2\pi)^{-1/2 n} \sigma^{-n}|H|^{-1/2} * exp\left\{-\frac{1}{2\sigma^2}(y-Xb)^T H^{-1} (y-Xb) \right\} \notag
\end{equation}\]</span></p>
<p>with <span class="math inline">\(var(y) = H*\sigma^2\)</span> and <span class="math inline">\(\theta^T = \left[\begin{array}{cc} b  &amp;  \sigma^2 \end{array} \right]\)</span></p>
</section>
<section id="maximization-of-likelihood" class="level2">
<h2 class="anchored" data-anchor-id="maximization-of-likelihood">Maximization of Likelihood</h2>
<ul>
<li>Set <span class="math inline">\(\lambda = log L\)</span></li>
<li>Compute partial derivatives of <span class="math inline">\(\lambda\)</span> with respect to all unknowns</li>
</ul>
<p><span class="math display">\[\frac{\partial \lambda}{\partial b}\]</span></p>
<p><span class="math display">\[\frac{\partial \lambda}{\partial \sigma^2}\]</span></p>
<ul>
<li>Set partial derivatives to <span class="math inline">\(0\)</span> and solve for unknowns</li>
<li>Use solutions as estimates</li>
</ul>
</section>
<section id="restricted-maximum-likelihood-reml" class="level2">
<h2 class="anchored" data-anchor-id="restricted-maximum-likelihood-reml">Restricted Maximum Likelihood (REML)</h2>
<ul>
<li>Problem with ML: estimate of <span class="math inline">\(\sigma^2\)</span> depends on <span class="math inline">\(b\)</span> <span class="math inline">\(\rightarrow\)</span> undesirable</li>
<li>Do transformations <span class="math inline">\(Sy\)</span> and <span class="math inline">\(Qy\)</span></li>
</ul>
<ol type="i">
<li>The matrix <span class="math inline">\(S\)</span> has rank <span class="math inline">\(n-t\)</span> and the matrix <span class="math inline">\(Q\)</span> has rank <span class="math inline">\(t\)</span></li>
<li>The result of the two transformations are independent, that means <span class="math inline">\(cov(Sy,Qy) = 0\)</span> which is met when <span class="math inline">\(SHQ^T = 0\)</span></li>
<li>The matrix S is chosen such that <span class="math inline">\(E(Sy) = 0\)</span> which means <span class="math inline">\(SX = 0\)</span></li>
<li>The matrix <span class="math inline">\(QX\)</span> is of rank <span class="math inline">\(t\)</span>, so that every linear function of the elements of <span class="math inline">\(Qy\)</span> estimate a linear function of <span class="math inline">\(b\)</span>.</li>
</ol>
</section>
<section id="reml-ii" class="level2">
<h2 class="anchored" data-anchor-id="reml-ii">REML II</h2>
<ul>
<li>From (i) and (ii) it follows that the likelihood <span class="math inline">\(L\)</span> of <span class="math inline">\(y\)</span> is the product of the likelihoods of <span class="math inline">\(Sy\)</span> (<span class="math inline">\(L^*\)</span>) and <span class="math inline">\(Qy\)</span> (<span class="math inline">\(L^{**}\)</span>) that means</li>
</ul>
<p><span class="math display">\[\lambda = \lambda^* + \lambda^{**}\]</span></p>
<ul>
<li>Variance components are estimated from <span class="math inline">\(\lambda^*\)</span> which will then be independent of <span class="math inline">\(b\)</span></li>
</ul>
</section>
<section id="bayesian-estimation" class="level2">
<h2 class="anchored" data-anchor-id="bayesian-estimation">Bayesian Estimation</h2>
<ul>
<li>Proposed already in the 80’s</li>
<li>Full implementation only in 1993</li>
<li>Requirements:
<ul>
<li>cheap computing and</li>
<li>good pseudo-random number generators</li>
</ul></li>
<li>Bayesian estimation is based on conditional posterior distribution of unknowns given the knowns</li>
<li>Conditional posterior distribution is computed from prior distribution of unknowns times the likelihood</li>
</ul>
</section>
<section id="model-1" class="level2">
<h2 class="anchored" data-anchor-id="model-1">Model</h2>
<ul>
<li>Univariate Gaussian linear mixed model</li>
</ul>
<p><span class="math display">\[y = Xb + Zu + e\]</span></p>
</section>
<section id="likelihood" class="level2">
<h2 class="anchored" data-anchor-id="likelihood">Likelihood</h2>
<ul>
<li>Data generating distribution</li>
</ul>
<p><span class="math display">\[y | b, u, \sigma_e^2 \sim \mathcal{N}(Xb + Zu, I * \sigma_e^2)\]</span></p>
<p>where <span class="math inline">\(I\)</span> is a <span class="math inline">\(n\times n\)</span> identity matrix and <span class="math inline">\(\sigma_e^2\)</span> is the variance of the random residuals.</p>
</section>
<section id="priors" class="level2">
<h2 class="anchored" data-anchor-id="priors">Priors</h2>
<ul>
<li>Prior distributions must be specified for all unknowns</li>
<li>Unknowns in our example are: <span class="math inline">\(b\)</span>, <span class="math inline">\(u\)</span>, <span class="math inline">\(\sigma_e^2\)</span> and <span class="math inline">\(\sigma_u^2\)</span></li>
<li>Prior distribution for
<ul>
<li><span class="math inline">\(b\)</span> is flat, i.e.&nbsp;<span class="math inline">\(p(b) \propto c\)</span></li>
<li><span class="math inline">\(u\)</span> Normal distribution as <span class="math inline">\(u | G, \sigma_u^2 \sim N(0, G*\sigma_u^2)\)</span></li>
<li><span class="math inline">\(\sigma_e^2\)</span> scaled inverse <span class="math inline">\(\chi^2\)</span>: <span class="math inline">\(p(\sigma_e^2 | \nu_e, s_e^2) \propto (\sigma_e^2)^{-\nu_e/2-1} exp(-{1\over 2}\nu_e s_e^2 / \sigma_e^2)\)</span></li>
<li><span class="math inline">\(\sigma_u^2\)</span> : <span class="math inline">\(p(\sigma_u^2 | \nu_u, s_u^2) \propto (\sigma_u^2)^{-\nu_u/2-1} exp(-{1\over 2}\nu_u s_u^2 / \sigma_u^2)\)</span></li>
</ul></li>
<li><span class="math inline">\(\nu_e\)</span>, <span class="math inline">\(\nu_s\)</span>, <span class="math inline">\(s_e^2\)</span> and <span class="math inline">\(s_u^2\)</span> are called hyper-parameters and must be determined</li>
</ul>
</section>
<section id="additional-terms" class="level2">
<h2 class="anchored" data-anchor-id="additional-terms">Additional Terms</h2>
<ul>
<li>Let</li>
</ul>
<p><span class="math display">\[\theta^T = (b^T, u^T) = (\theta_1, \theta_2, \ldots, \theta_N)\]</span></p>
<p><span class="math display">\[\theta_{-i} = (\theta_1, \theta_2, \ldots, \theta_{i-1}, \theta_{i+1}, \ldots, \theta_N)\]</span></p>
<ul>
<li>Further, let</li>
</ul>
<p><span class="math display">\[s^T = (s_u^2, s_e^2)\]</span> and</p>
<p><span class="math display">\[\nu^T = (\nu_u, \nu_e)\]</span></p>
</section>
<section id="joint-posterior-density" class="level2">
<h2 class="anchored" data-anchor-id="joint-posterior-density">Joint Posterior Density</h2>
<p>The joint posterior distribution can be written as</p>
<p><span class="math display">\[p(\theta, \sigma_u^2, \sigma_e^2 | y, s, \nu) \propto p(\theta) * p(\sigma_u^2 | \nu_u, s_u^2) * p(\sigma_e^2 | \nu_e, s_e^2) * p(y | \theta, \sigma_e^2)\]</span></p>
</section>
<section id="fully-conditional-posterior-densities-of-theta" class="level2">
<h2 class="anchored" data-anchor-id="fully-conditional-posterior-densities-of-theta">Fully Conditional Posterior Densities of <span class="math inline">\(\theta\)</span></h2>
<ul>
<li>Density of every single unknown component when setting all other components as known</li>
</ul>
<p><span class="math display">\[\theta_i | y, \theta_{-i}, \sigma_u^2, \sigma_e^2, s, \nu \sim \mathcal{N}(\tilde{\theta_i}, \tilde{v_i}) \]</span></p>
<p>where <span class="math inline">\(\tilde{\theta_i} = (r_i - \sum_{j=1,j\ne i}^N w_{ij}\theta_j)/w_{ii}\)</span> and <span class="math inline">\(\tilde{v_i} = \sigma_e^2 / w_{ii}\)</span>.</p>
<ul>
<li>vector <span class="math inline">\(r\)</span> is the vector of right-hand side of MME</li>
<li>matrix <span class="math inline">\(W\)</span> is the coefficient matrix of MME</li>
</ul>
</section>
<section id="fully-conditional-posterior-densities-of-sigma_e2" class="level2">
<h2 class="anchored" data-anchor-id="fully-conditional-posterior-densities-of-sigma_e2">Fully Conditional Posterior Densities of <span class="math inline">\(\sigma_e^2\)</span></h2>
<ul>
<li>scaled inverted chi-square distribution for <span class="math inline">\(\sigma_e^2\)</span></li>
</ul>
<p><span class="math display">\[\sigma_e^2 | y, \theta, \sigma_u^2, s, \nu \sim \tilde{\nu_e}\tilde{s_e}^2\chi_{\tilde{\nu_e}}^{-2}\]</span></p>
<ul>
<li>Parameters of the above distribution are defined as</li>
</ul>
<p><span class="math display">\[\tilde{\nu_e} = n + \nu_e\]</span> and</p>
<p><span class="math display">\[\tilde{s_e}^2 = \left[(y - Xb - Zu)^T(y - Xb - Zu) + \nu_e s_e^2 \right] / \tilde{\nu_e}\]</span></p>
</section>
<section id="fully-conditional-posterior-densities-of-sigma_u2" class="level2">
<h2 class="anchored" data-anchor-id="fully-conditional-posterior-densities-of-sigma_u2">Fully Conditional Posterior Densities of <span class="math inline">\(\sigma_u^2\)</span></h2>
<ul>
<li>scaled inverted chi-square distribution for <span class="math inline">\(\sigma_u^2\)</span></li>
</ul>
<p><span class="math display">\[\sigma_u^2 | y, \theta, \sigma_e^2, s, \nu \sim \tilde{\nu_u}\tilde{s_u}^2\chi_{\tilde{\nu_u}}^{-2}\]</span></p>
<ul>
<li>Parameters of the above distribution are defined as</li>
</ul>
<p><span class="math display">\[\tilde{\nu_u} = q + \nu_u\]</span> and</p>
<p><span class="math display">\[\tilde{s_u}^2 = \left[u^TG^{-1}u + \nu_u s_u^2 \right] / \tilde{\nu_u}\]</span></p>
</section>
<section id="implementation" class="level2">
<h2 class="anchored" data-anchor-id="implementation">Implementation</h2>
<ul>
<li>Step 1: set starting values for <span class="math inline">\(\theta\)</span>, <span class="math inline">\(\sigma_e^2\)</span> and <span class="math inline">\(\sigma_u^2\)</span></li>
<li>Step 2: draw random number for each component <span class="math inline">\(\theta_i\)</span> of <span class="math inline">\(\theta\)</span> from fully conditional distribution <span class="math inline">\(\mathcal{N}(\tilde{\theta_i}, \tilde{v_i})\)</span></li>
<li>Step 3: draw random number for <span class="math inline">\(\sigma_e^2\)</span> from <span class="math inline">\(\tilde{\nu_e}\tilde{s_e}^2\chi_{\tilde{\nu_e}}^{-2}\)</span></li>
<li>Step 4: draw random number for <span class="math inline">\(\sigma_u^2\)</span> from <span class="math inline">\(\tilde{\nu_u}\tilde{s_u}^2\chi_{\tilde{\nu_u}}^{-2}\)</span></li>
<li>Repeat steps 2-4 many times and store random numbers</li>
<li>Step 5: compute means of random numbers to get Bayesian estimates of unknowns <span class="math inline">\(\theta\)</span>, <span class="math inline">\(\sigma_e^2\)</span> and <span class="math inline">\(\sigma_u^2\)</span></li>
</ul>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>